### PR TITLE
net: lib: lwm2m_client_utils: Add support for connection pre-evaluation

### PR DESCRIPTION
**Pull request by juhaylinen**
Connection pre-evaluation estimates the LTE radio signal state and energy efficiency in a cell that the modem would use for data transmission. Application can use the information to decide when to start the data transmission.
If estimated energy usage is too high the application pauses the lwm2m client and waits for better signal state.
User can set evaluation period in milliseconds and maximum time in seconds to pause the transmission.

### CODE REVIEW

**Comment by juhaylinen** [comment]
This PR needs to wait for Zephyr upmerge to pass CI tests.

File: subsys/net/lib/lwm2m_client_utils/Kconfig [hunk 1]
```diff
@@ -248,6 +248,15 @@ config LWM2M_CLIENT_UTILS_RAI
 	  When AS RAI is configured, device may indicate that no further data is
 	  expected in the near future and the connection may be released.
 
+config LWM2M_CLIENT_UTILS_CONNEVAL
+	bool "Enable connection pre-evaluation"
```
Comment (rlubos) [inline subsys/net/lib/lwm2m_client_utils/Kconfig:252]: Please add `[EXPERIMENTAL]` suffix to the option description.

File: subsys/net/lib/lwm2m_client_utils/Kconfig [hunk 2]
```diff
@@ -248,6 +248,15 @@ config LWM2M_CLIENT_UTILS_RAI
 	  When AS RAI is configured, device may indicate that no further data is
 	  expected in the near future and the connection may be released.
 
+config LWM2M_CLIENT_UTILS_CONNEVAL
```
Comment (rlubos) [inline subsys/net/lib/lwm2m_client_utils/Kconfig:251]: I wonder if this option/API should include `LTE` keyword, like `LTE_CONNEVAL`. With other networking technologies on the table now, current name might be too generic.

**Comment by juhaylinen** [comment]
Refactored the code as discussed offline with @SeppoTakalo 

- Removed own work queue implementation and use delayed work instead.
- Use asynchronous AT command for connection pre-evaluation when evaluation is done multiple times. LTE link controller API may block too long for system work queue.

File: subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_conneval.c [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+#define AT_CONEVAL_RESULT_INDEX 1
+#define AT_CONEVAL_ENERGY_ESTIMATE_INDEX 3
+
+static struct k_work conneval_work;
+static struct k_work_delayable conneval_work_delayable;
+
+static struct lte_lc_conn_eval_params params;
+static enum lte_lc_energy_estimate energy_estimate;
+static uint64_t maximum_delay;
+static uint64_t poll_period;
```
Comment (VeijoPesonen) [inline subsys/net/lib/lwm2m_client_utils/lwm2m/lwm2m_conneval.c:33]:
```suggestion
static uint64_t poll_period; /* [ms] */
```

**Review by VeijoPesonen — CHANGES_REQUESTED** [review]
Nothing major, generally looks good to me.

**Review by VeijoPesonen — APPROVED** [review]
Looks good to me

**Review by VeijoPesonen — APPROVED** [review]

**Review by divipillai — APPROVED** [review]

**Review by SeppoTakalo — APPROVED** [review]

**Review by rlubos — APPROVED** [review]

**Pull request merged**